### PR TITLE
More accurate timestamps on risk card (EXPOSUREAPP-1991)

### LINKED ISSUE #948: CWA risk update timestamp different to Google timestamp (by one or two minutes)
## Avoid duplicates
* [X] Bug is not mentioned in the [FAQ](https://www.coronawarn.app/en/faq/)
* [X] Bug is specific for Android only, for general issues / questions that apply to iOS and Android please raise them in the [documentation repository](https://github.com/corona-warn-app/cwa-documentation)
* [X] Bug is not already reported in another issue

## Describe the bug

The last update time shown by CWA is sometimes one minute or two minutes earlier or later than the update time shown in the Google ENS UI.

## Expected behaviour

This is a minor issue.
The timestamp shown by CWA for the last update should be the same as the timestamp shown in the Google ENS UI in the overview screen and in the exposure checks log.

## Steps to reproduce the issue

1. View CWA risk status
Note text similar to "Updated: Today, 07:54"
2. Settings > Google > COVID-19 Exposure Notifications > three-dot icon > Exposure checks
Note latest timestamp in exposure checks list.

## Technical details

- Mobile device: Samsung Galaxy A5 (2017) SM-A520F
- Android version: 8.0.0
- CWA Version: ~~1.1.1~~ 2.4.2
- Google Exposure Notifications System Version: ~~15202902003~~ v1.8 (212116000)

Edit: Updated steps to reproduce for newer ENS UI, updated software versions and verified that the issue still occasionally occurs in current version of CWA and ENS.
 
---
Internal Tracking ID: [EXPOSUREAPP-1991](https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-1991)

### PR DESCRIPTION
**Pull request by d4rken**
Via #1473 we now have better infos about when the ENF calculation is finished. Let's display that timestamp instead of when we finished downloading the key pkgs (because download+provision is asynchronous from calculation).

This would fix #948 where the timestamp between the riskcard and the ENF log is off by 1-2 minutes due us using the download timestamp.

For now if there is not yet a timestamp available via `CalculationTracker` we'll fallback to the download timestamp otherwise everyone would see the wrong info after upgrading. We can remove that fallback in a few weeks after almost everyone has upgraded and `CalculationTracker` has a high chance of holding data to display.

## How to test
* Trigger a view key pkg submissions and compare the timestamp on the risk card with the timestamps from the ENF log.

### CODE REVIEW

**Comment by harambasicluka** [comment]
Hint for testing: Change the landscape to WRU-XD, it looks like that there aren't packages on INT.

**Review by harambasicluka — CHANGES_REQUESTED** [review]
Code looks fine.

Testing:
| Passed | UI    | Settings |
|--------|-------|----------|
| [x]    | 12:19 | 12:19    |
| [ ]    | 12:22 | 12:23    |
| [ ]    | 12:25 | 12:24    |
| [x]    | 12:30 | 12:30     |
| [ ]| 12:33 | 12:31 |

Tested on Samsung Galaxy S8 with Android 9 by time traveling from day to day.

```
2020-10-31 12:47:47.999 2157-2157/de.rki.coronawarnapp.test V/CalculationTrackerStorage: init()
2020-10-31 12:47:47.999 2157-2157/de.rki.coronawarnapp.test V/DefaultCalculationTracker: init()
2020-10-31 12:47:48.005 2157-2157/de.rki.coronawarnapp.test V/DefaultCalculationTracker:HD: init()
2020-10-31 12:47:48.011 2157-2216/de.rki.coronawarnapp.test V/DefaultCalculationTracker$calculationStates$2$setupAutoSave: Observing calculation changes.
2020-10-31 12:47:48.017 2157-2218/de.rki.coronawarnapp.test V/DefaultCalculationTracker:HD: internal onStart
2020-10-31 12:47:48.039 2157-2216/de.rki.coronawarnapp.test V/CalculationTrackerStorage: Loaded calculation data: {a36074a8-a2b3-42e2-b47b-f9a334647d19=Calculation(identifier=a36074a8-a2b3-42e2-b47b-f9a334647d19, startedAt=2020-10-31T11:43:13.290Z, result=NO_MATCHES, finishedAt=2020-10-31T11:43:30.186Z)}
2020-10-31 12:47:48.040 2157-2216/de.rki.coronawarnapp.test V/DefaultCalculationTracker:HD: startValue={a36074a8-a2b3-42e2-b47b-f9a334647d19=Calculation(identifier=a36074a8-a2b3-42e2-b47b-f9a334647d19, startedAt=2020-10-31T11:43:13.290Z, result=NO_MATCHES, finishedAt=2020-10-31T11:43:30.186Z)}
2020-10-31 12:47:48.042 2157-2219/de.rki.coronawarnapp.test V/CalculationTrackerStorage: Data didn't change, skipping save.
2020-10-31 12:47:48.042 2157-2216/de.rki.coronawarnapp.test V/DefaultCalculationTracker$calculationStates$2$setupTimeoutEnforcer: Running timeout check on: [Calculation(identifier=a36074a8-a2b3-42e2-b47b-f9a334647d19, startedAt=2020-10-31T11:43:13.290Z, result=NO_MATCHES, finishedAt=2020-10-31T11:43:30.186Z)]
2020-10-31 12:47:48.043 2157-2216/de.rki.coronawarnapp.test V/DefaultCalculationTracker$calculationStates$2$setupTimeoutEnforcer: Time now: 2020-10-31T11:47:48.042Z
2020-10-31 12:47:48.044 2157-2216/de.rki.coronawarnapp.test V/CalculationTrackerStorage: Data didn't change, skipping save.
2020-11-02 12:48:24.381 2157-2219/de.rki.coronawarnapp.test I/DefaultCalculationTracker: trackNewCalaculation(token=e1c204cd-892d-46e6-92e7-aa2e9b626555)
2020-11-02 12:48:24.385 2157-2222/de.rki.coronawarnapp.test V/CalculationTrackerStorage: Storing calculation data: {a36074a8-a2b3-42e2-b47b-f9a334647d19=Calculation(identifier=a36074a8-a2b3-42e2-b47b-f9a334647d19, startedAt=2020-10-31T11:43:13.290Z, result=NO_MATCHES, finishedAt=2020-10-31T11:43:30.186Z), e1c204cd-892d-46e6-92e7-aa2e9b626555=Calculation(identifier=e1c204cd-892d-46e6-92e7-aa2e9b626555, startedAt=2020-11-02T11:48:24.382Z, result=null, finishedAt=null)}
2020-11-02 12:48:43.512 2157-2223/de.rki.coronawarnapp.test I/DefaultCalculationTracker: finishCalculation(token=e1c204cd-892d-46e6-92e7-aa2e9b626555, result=NO_MATCHES)
2020-11-02 12:48:43.520 2157-2222/de.rki.coronawarnapp.test V/CalculationTrackerStorage: Storing calculation data: {a36074a8-a2b3-42e2-b47b-f9a334647d19=Calculation(identifier=a36074a8-a2b3-42e2-b47b-f9a334647d19, startedAt=2020-10-31T11:43:13.290Z, result=NO_MATCHES, finishedAt=2020-10-31T11:43:30.186Z), e1c204cd-892d-46e6-92e7-aa2e9b626555=Calculation(identifier=e1c204cd-892d-46e6-92e7-aa2e9b626555, startedAt=2020-11-02T11:48:24.382Z, result=NO_MATCHES, finishedAt=2020-11-02T11:48:43.514Z)}
2020-11-02 12:50:48.020 2157-2219/de.rki.coronawarnapp.test V/DefaultCalculationTracker$calculationStates$2$setupTimeoutEnforcer: Running timeout check on: [Calculation(identifier=a36074a8-a2b3-42e2-b47b-f9a334647d19, startedAt=2020-10-31T11:43:13.290Z, result=NO_MATCHES, finishedAt=2020-10-31T11:43:30.186Z), Calculation(identifier=e1c204cd-892d-46e6-92e7-aa2e9b626555, startedAt=2020-11-02T11:48:24.382Z, result=NO_MATCHES, finishedAt=2020-11-02T11:48:43.514Z)]
2020-11-02 12:50:48.025 2157-2219/de.rki.coronawarnapp.test V/DefaultCalculationTracker$calculationStates$2$setupTimeoutEnforcer: Time now: 2020-11-02T11:50:48.022Z
2020-11-02 12:50:48.030 2157-2219/de.rki.coronawarnapp.test V/CalculationTrackerStorage: Storing calculation data: {a36074a8-a2b3-42e2-b47b-f9a334647d19=Calculation(identifier=a36074a8-a2b3-42e2-b47b-f9a334647d19, startedAt=2020-10-31T11:43:13.290Z, result=NO_MATCHES, finishedAt=2020-10-31T11:43:30.186Z), e1c204cd-892d-46e6-92e7-aa2e9b626555=Calculation(identifier=e1c204cd-892d-46e6-92e7-aa2e9b626555, startedAt=2020-11-02T11:48:24.382Z, result=NO_MATCHES, finishedAt=2020-11-02T11:48:43.514Z)}
2020-11-04 12:51:08.258 2157-2219/de.rki.coronawarnapp.test I/DefaultCalculationTracker: trackNewCalaculation(token=b20454bf-1211-4a57-8cd7-cd1fdcb5bc62)
2020-11-04 12:51:08.262 2157-2222/de.rki.coronawarnapp.test V/CalculationTrackerStorage: Storing calculation data: {a36074a8-a2b3-42e2-b47b-f9a334647d19=Calculation(identifier=a36074a8-a2b3-42e2-b47b-f9a334647d19, startedAt=2020-10-31T11:43:13.290Z, result=NO_MATCHES, finishedAt=2020-10-31T11:43:30.186Z), e1c204cd-892d-46e6-92e7-aa2e9b626555=Calculation(identifier=e1c204cd-892d-46e6-92e7-aa2e9b626555, startedAt=2020-11-02T11:48:24.382Z, result=NO_MATCHES, finishedAt=2020-11-02T11:48:43.514Z), b20454bf-1211-4a57-8cd7-cd1fdcb5bc62=Calculation(identifier=b20454bf-1211-4a57-8cd7-cd1fdcb5bc62, startedAt=2020-11-04T11:51:08.259Z, result=null, finishedAt=null)}
2020-11-04 12:51:27.408 2157-2222/de.rki.coronawarnapp.test I/DefaultCalculationTracker: finishCalculation(token=b20454bf-1211-4a57-8cd7-cd1fdcb5bc62, result=NO_MATCHES)
2020-11-04 12:51:27.412 2157-2215/de.rki.coronawarnapp.test V/CalculationTrackerStorage: Storing calculation data: {a36074a8-a2b3-42e2-b47b-f9a334647d19=Calculation(identifier=a36074a8-a2b3-42e2-b47b-f9a334647d19, startedAt=2020-10-31T11:43:13.290Z, result=NO_MATCHES, finishedAt=2020-10-31T11:43:30.186Z), e1c204cd-892d-46e6-92e7-aa2e9b626555=Calculation(identifier=e1c204cd-892d-46e6-92e7-aa2e9b626555, startedAt=2020-11-02T11:48:24.382Z, result=NO_MATCHES, finishedAt=2020-11-02T11:48:43.514Z), b20454bf-1211-4a57-8cd7-cd1fdcb5bc62=Calculation(identifier=b20454bf-1211-4a57-8cd7-cd1fdcb5bc62, startedAt=2020-11-04T11:51:08.259Z, result=NO_MATCHES, finishedAt=2020-11-04T11:51:27.409Z)}
2020-11-04 12:53:48.018 2157-2218/de.rki.coronawarnapp.test V/DefaultCalculationTracker$calculationStates$2$setupTimeoutEnforcer: Running timeout check on: [Calculation(identifier=a36074a8-a2b3-42e2-b47b-f9a334647d19, startedAt=2020-10-31T11:43:13.290Z, result=NO_MATCHES, finishedAt=2020-10-31T11:43:30.186Z), Calculation(identifier=e1c204cd-892d-46e6-92e7-aa2e9b626555, startedAt=2020-11-02T11:48:24.382Z, result=NO_MATCHES, finishedAt=2020-11-02T11:48:43.514Z), Calculation(identifier=b20454bf-1211-4a57-8cd7-cd1fdcb5bc62, startedAt=2020-11-04T11:51:08.259Z, result=NO_MATCHES, finishedAt=2020-11-04T11:51:27.409Z)]
2020-11-04 12:53:48.020 2157-2218/de.rki.coronawarnapp.test V/DefaultCalculationTracker$calculationStates$2$setupTimeoutEnforcer: Time now: 2020-11-04T11:53:48.019Z
2020-11-04 12:53:48.027 2157-2218/de.rki.coronawarnapp.test V/CalculationTrackerStorage: Storing calculation data: {a36074a8-a2b3-42e2-b47b-f9a334647d19=Calculation(identifier=a36074a8-a2b3-42e2-b47b-f9a334647d19, startedAt=2020-10-31T11:43:13.290Z, result=NO_MATCHES, finishedAt=2020-10-31T11:43:30.186Z), e1c204cd-892d-46e6-92e7-aa2e9b626555=Calculation(identifier=e1c204cd-892d-46e6-92e7-aa2e9b626555, startedAt=2020-11-02T11:48:24.382Z, result=NO_MATCHES, finishedAt=2020-11-02T11:48:43.514Z), b20454bf-1211-4a57-8cd7-cd1fdcb5bc62=Calculation(identifier=b20454bf-1211-4a57-8cd7-cd1fdcb5bc62, startedAt=2020-11-04T11:51:08.259Z, result=NO_MATCHES, finishedAt=2020-11-04T11:51:27.409Z)}

```

**Comment by vaubaehn** [comment]
> For now if there is not yet a timestamp available via `CalculationTracker` we'll fallback to the download timestamp otherwise everyone would see the wrong info after upgrading. We can remove that fallback in a few weeks after almost everyone has upgraded and `CalculationTracker` has a high chance of holding data to display.

Wouldn't we need the fallback also later for new/first time installations?

**Comment by d4rken** [comment]
> Wouldn't we need the fallback also later for new/first time installations? Or after a data reset?

At that point the user wouldn't expect a time stamp though, so displaying the text for "no value being available" is okay.

I think it's wouldn't even come to that, as it would display the grey card  for "not enough data" which doesn't show a timestamp. Right? :thinking:

**Comment by d4rken** [comment]
Marked as draft as it turns out that the timestamp from Google does not fit the expected pattern. Going to clarify with Google what their timestamp actually represents, from Luka's tests, it's not just rounding up/down either.

**Comment by vaubaehn** [comment]
> At that point the user wouldn't expect a time stamp though, so displaying the text for "no value being available" is okay.
> 
> I think it's wouldn't even come to that, as it would display the grey card for "not enough data" which doesn't show a timestamp. Right? :thinking:

I agree, if the risk card refresh is triggered by receiving the two 'exposure check fnished' broadcasts from ENF only ('unknown risk'  until the first exposure check result is retrieved), not by closing ProvideDiagnosisKeysTransaction anymore... In my age, the brain needs more effort to re-connect changed flows consistently ;)

**Comment by vaubaehn** [comment]
> from Luka's tests, it's not just rounding up/down either.

@harambasicluka 

Hi Luka, if you find some time, could you help me understand your tests:
Table: UI is CWA and Settings is ENF?
Log: the log doesn't correspond to the table, just shows that logging works in general, right?

What I also don't understand:
[Corona-Warn-App/src/test/java/de/rki/coronawarnapp/ui/tracing/card/TracingCardStateTest.kt](https://github.com/corona-warn-app/cwa-app-android/pull/1487/files#diff-79aaac5b9c3afd793f7cfe4f50e202bf809df6c94277542abd564e398caa1ef8) - line 58
Corona-Warn-App/src/test/java/de/rki/coronawarnapp/ui/tracing/common/BaseTracingStateTest.kt - line 65
Corona-Warn-App/src/test/java/de/rki/coronawarnapp/ui/tracing/details/TracingDetailsStateTest.kt - line 60
here `lastENFCalculation` is always referring to `lastTimeDiagnosisKeysFetched` - shouldn't it refer to `lastUpdateDate` also for the tests?

Thanks in advance!

**Comment by d4rken** [comment]
> What I also don't understand:
Corona-Warn-App/src/test/java/de/rki/coronawarnapp/ui/tracing/card/TracingCardStateTest.kt - line 58
Corona-Warn-App/src/test/java/de/rki/coronawarnapp/ui/tracing/common/BaseTracingStateTest.kt - line 65
Corona-Warn-App/src/test/java/de/rki/coronawarnapp/ui/tracing/details/TracingDetailsStateTest.kt - line 60
here lastENFCalculation is always referring to lastTimeDiagnosisKeysFetched - shouldn't it refer to lastUpdateDate also for the tests?

It should. Doesn't change the results though, just local variable names. Will fix before this gets merged, (if it get's merged), after we get a response from Google. Good catch :+1:

**Comment by harambasicluka** [comment]
> > from Luka's tests, it's not just rounding up/down either.
> 
> @harambasicluka
> 
> Hi Luka, if you find some time, could you help me understand your tests:
> Table: UI is CWA and Settings is ENF?
> Log: the log doesn't correspond to the table, just shows that logging works in general, right?

@vaubaehn I just checked the timestamp displayed on the risk card and than went in to the os settings (Settings > Google > COVID-19 exposure notifications > Exposure checks) and compared the time displayed here.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/tracing/card/TracingCardStateProvider.kt [hunk 1]
```diff
@@ -46,8 +49,8 @@ class TracingCardStateProvider @Inject constructor(
         tracingRepository.activeTracingDaysInRetentionPeriod.onEach {
             Timber.v("activeTracingDaysInRetentionPeriod: $it")
         },
-        tracingRepository.lastTimeDiagnosisKeysFetched.onEach {
-            Timber.v("lastTimeDiagnosisKeysFetched: $it")
+        enfClient.latestFinishedCalculation().onEach {
+            Timber.v("latestFinishedCalculation: $it")
```
Comment (chris-cwa) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/tracing/card/TracingCardStateProvider.kt:53]: https://github.com/corona-warn-app/cwa-app-android/pull/1412#discussion_r512516696

**Comment by d4rken** [comment]
This PR is now too stale, still waiting on clarification from Google.

**Pull request closed without merging**